pramen/core/src/main/scala/za/co/absa/pramen/core/runner/task/TaskRunnerBase.scala

```diff
@@ -115,7 +118,30 @@ abstract class TaskRunnerBase(conf: Config,
   /** Runs a task in the single thread. Performs all task logging and notification sending activities. */
   protected def runTask(task: Task): RunStatus = {
     val started = Instant.now()
+    task.job.operation.killMaxExecutionTimeSeconds match {
+      case Some(timeout) =>
+        var runStatus: RunStatus = null
+
+        try {
+          ThreadUtils.runWithTimeout(Duration(timeout, TimeUnit.SECONDS)) {
+            log.info(s"Running ${task.job.name} with the hard timeout = $timeout seconds.")
+            val status = doValidateAndRunTask(task)
+            this.synchronized {
```

Comment:
I didn't fully get the point why you need a synchronized keyword here. 
`runStatus` is a local variable. If two threads call the root method, two different variables will be allocated.

Reply:
Everything inside `runWithTimeout {}` runs in a different thread. `runStatus` reference is passed to that thread, effectively making it shared. Since `runStatus` is going to be updated from another thread, we need memory fences that `synchronize` block provides